File: java-proxy/src/main/java/com/proxy/AuthManager.java
package com.proxy;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Instant;
import java.util.concurrent.atomic.AtomicReference;
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class AuthManager {
    private final Config config;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper = new ObjectMapper();

    private final AtomicReference<String> masterToken = new AtomicReference<>();
    private final ReentrantReadWriteLock tokenLock = new ReentrantReadWriteLock();
    private final ReentrantLock reauthMutex = new ReentrantLock();
    private volatile Instant lastReauthAttempt = Instant.MIN;

    public AuthManager(Config config, HttpClient httpClient) {
        this.config = config;
        this.httpClient = httpClient;
    }

    public String getToken() {
        tokenLock.readLock().lock();
        try {
            return masterToken.get();
        } finally {
            tokenLock.readLock().unlock();
        }
    }

    public void performInitialAuthentication() throws InterruptedException {
        boolean authenticated = false;
        String token = null;

        for (int i = 0; i < config.getMaxRetries(); i++) {
            Logger.log(Logger.Level.INFO, "Attempting to authenticate", String.format("(Attempt %d/%d)", i + 1, config.getMaxRetries()));
            try {
                token = authenticate();
                masterToken.set(token);
                authenticated = true;
                Logger.log(Logger.Level.SUCCESS, "Authenticated and retrieved initial token.");
                break;
            } catch (Exception e) {
                Logger.log(Logger.Level.ERROR, "Authentication failed!", e.getMessage());
                if (i < config.getMaxRetries() - 1) {
                    Logger.log(Logger.Level.INFO, String.format("Retrying in %dms...", config.getRetryIntervalMs()));
                    Thread.sleep(config.getRetryIntervalMs());
                }
            }
        }

        if (!authenticated) {
            Logger.log(Logger.Level.ERROR, "Could not authenticate after multiple retries. Exiting.");
            throw new RuntimeException("Could not authenticate with target server after multiple retries. Exiting.");
        }
    }

    public void refreshTokenIfNeeded() {
        if (!reauthMutex.tryLock()) {
            return;
        }
        try {
            if (Instant.now().isBefore(lastReauthAttempt.plusMillis(config.getReauthCooldownMs()))) {
                Logger.log(Logger.Level.INFO, "Re-authentication cooldown active. Please wait.");
                return;
            }
            lastReauthAttempt = Instant.now();
            Logger.log(Logger.Level.INFO, "Attempting to refresh token in background.");
            String newToken = authenticate();

            tokenLock.writeLock().lock();
            try {
                masterToken.set(newToken);
                Logger.log(Logger.Level.SUCCESS, "Successfully refreshed master token.");
            } finally {
                tokenLock.writeLock().unlock();
            }
        } catch (Exception e) {
            Logger.log(Logger.Level.ERROR, "Failed to refresh token", e);
        } finally {
            reauthMutex.unlock();
        }
    }

    private String authenticate() throws IOException, InterruptedException {
        String authUrl = "https://" + config.getTargetHost() + ":" + config.getTargetPort() + "/api/v1/auth";
        Logger.log(Logger.Level.INFO, "Posting to auth endpoint", "URL: " + authUrl);

        AuthRequest authRequest = new AuthRequest(config.getClientId(), config.getClientSecret());
        String requestBody = objectMapper.writeValueAsString(authRequest);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(authUrl))
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(requestBody))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 200) {
            Logger.log(Logger.Level.WARN, "Authentication failed with non-OK status.", String.format("Status: %d %s", response.statusCode(), HttpStatus.getStatusText(response.statusCode())));
            throw new AuthenticationException("auth failed", request.method(), authUrl, response.statusCode(), response.body());
        }

        AuthResponse authResponse = objectMapper.readValue(response.body(), AuthResponse.class);
        return authResponse.master_token();
    }

    public static class AuthenticationException extends IOException {
        private final String operation;
        private final String url;
        private final int statusCode;
        private final String responseBody;

        public AuthenticationException(String message, String operation, String url, int statusCode, String responseBody) {
            super(String.format("%s \"%s\": address %s : %s", operation, url, responseBody, message));
            
            this.operation = operation;
            this.url = url;
            this.statusCode = statusCode;
            this.responseBody = responseBody;
        }

        public String getOperation() {
            return operation;
        }

        public String getUrl() {
            return url;
        }

        public int getStatusCode() {
            return statusCode;
        }

        public String getResponseBody() {
            return responseBody;
        }
    }
}

File: node-proxy/src/auth-manager.ts
import https from 'https';
import { sharedHttpsAgent } from './agent';
import { Config } from './config';
import { logger, LogLevel } from './logger';

export class AuthManager {
  private masterToken: string = '';
  private isReauthenticating = false;
  private lastReauthAttempt = 0;

  constructor(private cfg: Config) {}

  public getToken(): string {
    return this.masterToken;
  }

  public async initialAuthenticate(): Promise<void> {
    for (let i = 0; i < this.cfg.maxRetries; i++) {
      try {
        logger(LogLevel.INFO, `Attempting to authenticate`, `(Attempt ${i + 1}/${this.cfg.maxRetries})`);
        this.masterToken = await this.performAuthRequest();
        logger(LogLevel.SUCCESS, 'Authenticated and retrieved initial token.');
        return;
      } catch (err: any) {
        logger(LogLevel.ERROR, 'Authentication failed!', err.message);
        if (i < this.cfg.maxRetries - 1) {
          logger(LogLevel.INFO, `Retrying in ${this.cfg.retryIntervalMs}ms...`);
          await new Promise(resolve => setTimeout(resolve, this.cfg.retryIntervalMs));
        } else {
          logger(LogLevel.ERROR, `Could not authenticate after multiple retries. Exiting.`);
          throw new Error('Could not authenticate with target server after multiple retries. Exiting.');
        }
      }
    }
  }

  public handleForbiddenStatus(): void {
    logger(LogLevel.WARN, 'Received 403 Forbidden from target. Triggering re-authentication.');
    if (this.isReauthenticating) return;

    const now = Date.now();
    if (now - this.lastReauthAttempt < this.cfg.reauthCooldownMs) {
      logger(LogLevel.INFO, 'Re-authentication cooldown active. Please wait.');
      return;
    }

    this.isReauthenticating = true;
    this.lastReauthAttempt = now;

    logger(LogLevel.INFO, 'Attempting to refresh token in background.');
    this.performAuthRequest()
      .then(newToken => {
        this.masterToken = newToken;
        logger(LogLevel.SUCCESS, 'Successfully refreshed master token in background.');
      })
      .catch(err => {
        logger(LogLevel.ERROR, 'Failed to refresh token', err.message);
      })
      .finally(() => {
        this.isReauthenticating = false;
      });
  }

  private performAuthRequest(): Promise<string> {
    const authUrl = `${this.cfg.targetHost}:${this.cfg.targetPort}/api/v1/auth`;
    logger(LogLevel.INFO, 'Posting to auth endpoint', `URL: ${authUrl}`);

    return new Promise((resolve, reject) => {
      const postData = JSON.stringify({
        client_id: this.cfg.clientId,
        client_secret: this.cfg.clientSecret,
      });

      const options = {
        hostname: this.cfg.targetHost,
        port: this.cfg.targetPort,
        path: '/api/v1/auth',
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(postData),
        },
        agent: sharedHttpsAgent,
      };

      const req = https.request(options, res => {
        let data = '';
        res.on('data', chunk => (data += chunk));
        res.on('end', () => {
          if (res.statusCode !== 200) {
            const statusText = res.statusMessage || '';
            logger(LogLevel.WARN, `Authentication failed with non-OK status. Status: ${res.statusCode} ${statusText}`);
            return reject(new Error(`${options.method} "${authUrl}": address ${data}: auth failed`));
          }
          try {
            resolve(JSON.parse(data).master_token);
          } catch (e) {
            reject(new Error('Failed to parse auth response JSON.'));
          }
        });
      });
      req.on('error', e => reject(new Error(`Auth request failed: ${e.message}`)));
      req.write(postData);
      req.end();
    });
  }
}
